Add debug logs when pod is deleted by the operator

adamantal opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | no
| API breaks?     | no
| Deprecations?   | no
| Related tickets | -
| License         | Apache 2.0

### What's in this PR?
<!-- Explain the contents of the PR. Give an overview about the implementation, which decisions were made and why. -->
A few lines of logs have been added in this PR to trace when a kafka broker pod (and service, PVC etc.) is deleted by the operator.

### Why?
<!-- Which problem does the PR fix? (Please remove this section if you linked an issue above) -->
Currently the operator does not display much info, if it deletes a Broker *on purpose*. That is what this PR intended to change.

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline

stoader commented on pkg/resources/kafka/kafka.go:
```diff
@@ -384,6 +384,7 @@ OUTERLOOP:
 			if err != nil {
 				return errors.WrapIfWithDetails(err, "could not delete broker", "id", broker.Labels["brokerId"])
 			}
+			log.Info(fmt.Sprintf("Broker %s deleted", broker.Labels["brokerId"]))
```
Use structured logging e.g.
```
log.Info("deleting broker", "brokerId",  broker.Labels["brokerId"], "namespace", broker.GetNamespace(), "pod", broker.GetName())
```

Also log pod deletion here :https://github.com/banzaicloud/kafka-operator/blob/master/pkg/resources/kafka/kafka.go#L736

adamantal commented:
Thanks for the review @stoader. Updated the PR based on your review.

I added structured logging, though the `namespace` field was redundant due to previous `log.WithValues` calls. 
Example log entry (prettified):
```json
{
  "level": "info",
  "ts": "2020-10-26T13:32:09.293Z",
  "logger": "controllers.KafkaCluster",
  "msg": "broker pod deleted",
  "Request.Namespace": "kafka/kafka",
  "Request.Name": "kafka",
  "component": "kafka",
  "clusterName": "kafka",
  "clusterNamespace": "kafka",
  "brokerId": "3",
  "pod": "kafka-3-6gjmx"
}
```

Test cases performed (locally on docker-desktop):
- Upscaled to 4, then downscaled to 3 brokers. The log entry is present in the operator logs along with the CC tasks and other details.
- Changed one of the ConfigMap of the brokers. The broker got deleted and then recreated by the operator and the corresponding log entry ("broker pod deleted") is present in the logs. Actually I've found two occasions of this log, one with false and one with true `hasTerminatedContainers` field. I added that field in order to make it clear how code had flowed before the deletion.

Removed the WIP/draft flag from the PR. Feel free to add any comment you have.

hi-im-aren commented on pkg/resources/kafka/kafka.go:
```diff
@@ -384,29 +384,37 @@ OUTERLOOP:
+					pvcName := volume.PersistentVolumeClaim.ClaimName
 					err = r.Client.Delete(context.TODO(), &corev1.PersistentVolumeClaim{ObjectMeta: metav1.ObjectMeta{
-						Name:      volume.PersistentVolumeClaim.ClaimName,
+						Name:      pvcName,
```
Was ` Name:  volume.PersistentVolumeClaim.ClaimName,`  not straightforward? 
I agree with the configmap and servicename refactors above because the fmt.Sprintf() made them long, but not sure about this one here.

adamantal replied:
I thought it's a bit easier to read, but I reverted this part.

stoader commented on pkg/resources/kafka/kafka.go:
```diff
@@ -384,23 +384,30 @@ OUTERLOOP:
 					return errors.WrapIfWithDetails(err, "could not delete configmap for broker", "id", broker.Labels["brokerId"])
 				}
+			} else {
+				log.V(1).Info("configMap for broker deleted", "configMap name", configMapName, "brokerId", broker.Labels["brokerId"])
```
```suggestion
				log.V(1).Info("configmap for broker deleted", "configmap name", configMapName, "brokerId", broker.Labels["brokerId"])
```

stoader commented on pkg/resources/kafka/kafka.go:
```diff
@@ -737,6 +746,7 @@ func (r *Reconciler) reconcileKafkaPod(log logr.Logger, desiredPod *corev1.Pod)
 		if err != nil {
 			return errorfactory.New(errorfactory.APIFailure{}, err, "deleting resource failed", "kind", desiredType)
 		}
+		log.Info("broker pod deleted", "pod", currentPod.ObjectMeta.Name, "brokerId", currentPod.Labels["brokerId"], "hasTerminatedContainers", k8sutil.IsPodContainsTerminatedContainer(currentPod))
```
```suggestion
		log.Info("broker pod deleted", "pod", currentPod.GetName(), "brokerId", currentPod.Labels["brokerId"], "hasTerminatedContainers", k8sutil.IsPodContainsTerminatedContainer(currentPod))
```

adamantal commented:
Added some extra logging with the `ExitCode` and `Reason` fields for the terminated containers.
Tested locally, provides the following log entry in development settings:
```
...
2020-10-27T11:40:43.076Z	INFO	controllers.KafkaCluster	terminated container for broker pod	{"Request.Namespace": "kafka/kafka", "Request.Name": "kafka", "component": "kafka", "clusterName": "kafka", "clusterNamespace": "kafka", "kind": "*v1.Pod", "pod": "kafka-0-nd4nc", "brokerId": "0", "containerName": "kafka", "exitCode": 0, "reason": "Completed"}
...
```

baluchicken reviewed (approved):
LGTM
